	SuSE43041 - LTC10011
	PCI Hotplug Remove can cause slab corruption

===== arch/ppc64/kernel/eeh.c 1.28 vs edited =====
--- 1.28/arch/ppc64/kernel/eeh.c	Mon Jul 12 18:29:16 2004
+++ edited/arch/ppc64/kernel/eeh.c	Wed Jul 14 15:40:47 2004
@@ -250,6 +250,7 @@
 static void __pci_addr_cache_insert_device(struct pci_dev *dev)
 {
 	struct device_node *dn;
+	int really_did_insert = 0;
 	int i;
 
 	dn = pci_device_to_OF_node(dev);
@@ -268,7 +269,6 @@
 #endif
 		return;
 	}
-	pci_dev_get(dev);
 
 	/* Walk resources on this device, poke them into the tree */
 	for (i = 0; i < DEVICE_COUNT_RESOURCE; i++) {
@@ -282,6 +282,11 @@
 		if (start == 0 || ~start == 0 || end == 0 || ~end == 0)
 			 continue;
 		pci_addr_cache_insert(dev, start, end, flags);
+		really_did_insert = 1;
+	}
+
+	if (really_did_insert) {
+		pci_dev_get (dev);
 	}
 }
 
@@ -305,6 +310,7 @@
 static inline void __pci_addr_cache_remove_device(struct pci_dev *dev)
 {
 	struct rb_node *n;
+	int really_did_remove = 0;
 
 restart:
 	n = rb_first(&pci_io_addr_cache_root.rb_root);
@@ -315,11 +321,14 @@
 		if (piar->pcidev == dev) {
 			rb_erase(n, &pci_io_addr_cache_root.rb_root);
 			kfree(piar);
+			really_did_remove = 1;
 			goto restart;
 		}
 		n = rb_next(n);
 	}
-	pci_dev_put(dev);
+	if (really_did_remove) {
+	   pci_dev_put(dev);
+	}
 }
 
 /**
